### ISSUE TITLE
DNS resolution is unstable

### ISSUE THREAD

**Issue by nertpinx** [alpinelinux/docker-alpine#203]
One broken nameserver in `/etc/resolv.conf` can break (almost) all DNS translations, no matter what the order is.

```
podman run --rm -it docker.io/library/alpine:edge
/ # echo -e "nameserver 10.0.2.3\nnameserver 0.0.0.1" >/etc/resolv.conf
/ # cat /etc/resolv.conf
nameserver 10.0.2.3
nameserver 0.0.0.1
/ # /usr/bin/time -f %e getent hosts alpinelinux.org
Command exited with non-zero status 2
10.00
/ # nslookup alpinelinux.org
Server:		10.0.2.3
Address:	10.0.2.3:53

Non-authoritative answer:
Name:	alpinelinux.org
Address: 147.75.101.119

Non-authoritative answer:
Name:	alpinelinux.org
Address: 2604:1380:2002:4001::4

/ # echo -e "nameserver 10.0.2.3" >/etc/resolv.conf
/ # cat /etc/resolv.conf
nameserver 10.0.2.3
/ # /usr/bin/time -f %e getent hosts alpinelinux.org
2604:1380:2002:4001::4  alpinelinux.org  alpinelinux.org
0.00
/ # 
```

As you can see the 0.0.0.1 (chosen to be an invalid DNS server) can break the DNS resolution even if the other server is perfectly valid. In any case `nslookup` is still working, but that does not help with other programs who are (I am guessing) using just `getaddrinfo()`.

Per https://tldp.org/LDP/nag/node84.html the nameservers should be tried in order. That means when the first nameserver times out the second one should be tried, or all of them can be tried and first reply used, but the second one timing out should not break the first one. 

This breaks on edge, 3.14, 3.13, 3.10, 3.9, but works on non-alpine images like fedora or debian (various versions). I am suspecting a musl bug, but I do not have an easy way of testing that itself. Let me know if I should file this issue elsewhere. Thank you.

**Comment by nertpinx** [comment]
I did some test with various musl versions and musl itself does not exhibit this behaviour (the host `/etc/resolv.conf` is adjusted properly, the binary is fully statically linked) *until* I try it in the container. Then it fails again. So it cannot just be something in musl itself, it has to be something with either alpine linux or this container.

**Comment by ncopa** [comment]
Musl resolver will not try them in order, it uses all listed nameservers in parallel as described in https://wiki.musl-libc.org/functional-differences-from-glibc.html#Name-Resolver/DNS)

But it does not sound right that it ends up with a failure when there is a non-responding server. @richfelker any ideas what is going on here?

**Comment by richfelker** [comment]
Indeed I cannot reproduce this. After creating and binding a local port for the UDP socket to send queries from, there is no code in the DNS transport backend that errors out on any kind of network or syscall errors. It keeps going until it gets a conclusive response (success or nxdomain) or the timeout is hit.

If it's exiting early in your container environment, then either 0.0.0.1 is getting routed to some nameserver that's giving a conclusive negative response for the query, or something to do with your configuration is making the operation fail before it starts. For example if you've provided an example config different from your actual config, and your actual config has an IPv6 nameserver, and your container has seccomp rules that disallow IPv6 with the wrong errno codes (anything other than EAFNOSUPPORT) then it could cause the operation to fail.

If you want to investigate this further, run the failing command under strace and post the strace log.

**Comment by ncopa** [comment]
I am not able to reproduce either @nertpinx can you please try `apk add strace; strace getent hosts alpinelinux.org`. Might need to run the container with `--privileged`.

**Comment by nertpinx** [comment]
So I tried various things and I am even more confused now than before.
Running `strace getent hosts alpinelinux.org` did not show anything weird, so I tried it with an output file in order not to mess with the output and it actually works and returns the proper info:
```
/ # strace -o getent.txt getent hosts alpinelinux.org
147.75.101.119    alpinelinux.org  alpinelinux.org
```
So I tried it with `strace ping -c1 alpinelinux.org` next and it again looked like it is working. However when running with `-o` this time it actually resulted in an error:
```
/ # strace -o ping.txt ping -c1 alpinelinux.org
ping: bad address 'alpinelinux.org'
```
So I guess the `ping.txt` will, in this case, be more informative than `getent.txt`, so I am attaching that here: [ping.txt](https://github.com/alpinelinux/docker-alpine/files/7186649/ping.txt)

To give you more explanation for how this happens, in order to get this more easily reproduced, the whole reasoning behind this is that the container usually has `nameserver 8.8.8.8` appended in `/etc/resolv.conf` by default and I am in a network where DNS requests outside of that network are prohibited for various (valid) reasons. My guess is that those packets are just dropped and that's why there is no response to those. Hope that helps with the reproducer as it works with *some* addresses that should be bogus, I guess `0.0.0.1` was a bad choice on my side, sorry for that.

What's also really weird is that if I run the lookups in a loop overnight then couple of them succeed (e.g. two). I disregard that as a race error or similar.

**Comment by richfelker** [comment]
OK, the strace output is showing the lookup times out. It does not seem to have anything to do with your original report about a bad nameserver in the list. Neither nameserver (10.0.2.3 nor 8.8.8.8) is responding to the AAAA (IPv6) address lookup, and therefore the final result is a timeout. Presumably you have some firewall/redirection of DNS traffic to a broken nameserver that's trying to block IPv6 lookups or similar.

**Comment by richfelker** [comment]
Are you by any chance using Docker for Mac? It at least used to have lots of problems stemming from intercepting and rewriting DNS requests. See for example https://github.com/docker/for-mac/issues/5020. I'm not sure what the status on fix in releases is.

**Comment by richfelker** [comment]
I don't think I read your comment sufficiently and missed something that may be important. When you say:

> DNS requests outside of that network are prohibited for various (valid) reasons

Do you mean requests to other nameservers than 8.8.8.8 are prohibited? If so, it sounds like you may have some sort of active firewall rule that's blocking the source port as soon as it sees the request to a prohibited destination address, which ends up also blocking further queries to 8.8.8.8 (made from the same port). This would explain the rare timing-dependent event of some lookups working.

If this is the case, my advice would be to replace it with a stateless rule that blocks only the individual packets to prohibited addresses rather than persistent rules for the port. But if that's hard to change it would suffice just not to add prohibited addresses to your resolv.conf.

**Comment by nertpinx** [comment]
> I don't think I read your comment sufficiently and missed something that may be important. When you say:
> 
> > DNS requests outside of that network are prohibited for various (valid) reasons
> 
> Do you mean requests to other nameservers than 8.8.8.8 are prohibited? If so, it sounds like you may have some sort of active firewall rule that's blocking the source port as soon as it sees the request to a prohibited destination address, which ends up also blocking further queries to 8.8.8.8 (made from the same port). This would explain the rare timing-dependent event of some lookups working.
> 

Not that way, every DNS request outside the local network is blocked, so `8.8.8.8` is the one that is blocked in this case. That nameserver is added by default by podman, docker does something similar IIRC.

> If this is the case, my advice would be to replace it with a stateless rule that blocks only the individual packets to prohibited addresses rather than persistent rules for the port. But if that's hard to change it would suffice just not to add prohibited addresses to your resolv.conf.

I do not see into whether that rule is based on a port (but I guess that has to be), as I am not in control of that network.

Anyway, I managed to capture both success and a failure with strace (with the same config) of ping, which tells me there is an issue with the request timing out **sometimes**, which is even weirder. The only thing I can think of is that sending multiple queries would fill some buffer, but at the same time that does not make sense to me. But you are right that whenever the working nameserver returns both ipv6 and ipv4 replies then the non-working nameservers do not affect the functionality at all. I wonder where and how to search next.

**Comment by richfelker** [comment]
I think it's almost certain you have an active firewall or "transparent" proxy that's doing something wrong with DNS traffic. Perhaps it's intercepting and rewriting all DNS query packets to go to the local nameserver rather than blocking the forbidden addresses, and then the local nameserver is something that gets confused when it sees two identical queries.

Does the issue reliably go away if you take out the blocked nameserver from `resolv.conf`?

**Comment by nertpinx** [comment]
Not sure how "wrong" that could be, to my knowledge it only drops DNS traffic going outside of the network.

Yes the issue goes away if I remove the blocked nameservers from resolv.conf. If that was not the case I would be searching for an issue in my own setup.

And I know you did not suggest the following (yet?), but just so we are on the same page, possibly with other participants, because those are automatically added by various container solutions/runtimes, "just removing them" is not an easy, automatic solution. I would rather figure out the real issue here.

Last thing I tried was reproducing all three behaviours (known-working solution, the broken solution, and the occasional working output when run in the "broken" environment, which happens very often when running under strace). For all of these I captured the strace output and the packets that were used with tcpdump, which can be opened in wireshark, for example. When looking at the data bear in mind that 10.0.2.3 is mapped to the localhost address in the containers' slirp stack. And yes, that slirp was the first thing I blamed before even thinking of creating this issue.

Since GitHub does not support attaching pcap files (or a tar file for that matter), I am attaching a zip file with all the captures to be examined here: [alpine-issue.zip](https://github.com/alpinelinux/docker-alpine/files/7202916/alpine-issue.zip)

Could that be an issue with `poll()` there?

**Comment by richfelker** [comment]
Mentioning slirp a long time ago would have been really helpful. The problem is almost surely related to it. I take it your pcaps are taken from outside the container, not inside it, since we're seeing 127.0.0.1 not the address the process in the container is querying?

**Comment by nertpinx** [comment]
Sure I can. And I am sorry if my comment sounded like blaming just the container. So here are the outputs: [alpine-issue.zip](https://github.com/alpinelinux/docker-alpine/files/7203972/alpine-issue.zip)

You are right that I only see the replies in the working case, however when checking the "weird" case (described above) there is no reply either, but it worked.

**Comment by nertpinx** [comment]
Even before reading your last reply I realised I could do very easy test to see if libslirp is causing the issue here. I was blaming slirp4netns first, but I now tested the alpine image in a VM where I could easily switch between usermode networking (using libslirp) and a virtualised one. Of course it broke with the slirp one and I should have tested this way before.

Anyway, you make a good point, however the 8.8.8.8 address does not get remapped to 127.0.0.1, it should stay 8.8.8.8 and so slirp should be able to distinguish between the packet flows. However it is now easily proven that libslirp is the issue since it is the only thing that needs changing in order for this to get broken.

Thank you for your time and sorry for not checking even more thoroughly before (although I already spent way too much time on this).

Have a nice day ;)
